File: netlify/functions/spotify_current.js
// These variables will act as an in-memory cache.
// A "warm" function invocation can reuse the token from a previous run.
let accessToken = null;
let tokenExpiresAt = 0;

async function refreshAccessToken() {
    const {
        SPOTIFY_CLIENT_ID,
        SPOTIFY_CLIENT_SECRET,
        SPOTIFY_REFRESH_TOKEN
    } = process.env;

    // In Node.js, we use Buffer to create a Base64 string.
    const authHeader = Buffer.from(`${SPOTIFY_CLIENT_ID}:${SPOTIFY_CLIENT_SECRET}`).toString('base64');

    // To send 'application/x-www-form-urlencoded' data, we use URLSearchParams.
    const body = new URLSearchParams({
        grant_type: 'refresh_token',
        refresh_token: SPOTIFY_REFRESH_TOKEN,
    });

    const response = await fetch('https://accounts.spotify.com/api/token', {
        method: 'POST',
        headers: {
            'Authorization': `Basic ${authHeader}`,
            'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: body.toString(),
    });

    const data = await response.json();

    if (!response.ok) {
        throw new Error(`Failed to refresh token: ${JSON.stringify(data)}`);
    }

    accessToken = data.access_token;
    // Date.now() gives milliseconds, so divide by 1000 to get seconds like Python's time.time().
    // We subtract 60 seconds as a buffer to ensure the token is still valid.
    tokenExpiresAt = (Date.now() / 1000) + data.expires_in - 60;
}

async function getCurrentPlaying() {
    // Check if the token is missing or has expired.
    if (!accessToken || (Date.now() / 1000) >= tokenExpiresAt) {
        await refreshAccessToken();
    }

    const response = await fetch('https://api.spotify.com/v1/me/player/currently-playing', {
        headers: {
            'Authorization': `Bearer ${accessToken}`,
        },
    });

    // 204 No Content: Nothing is playing.
    if (response.status === 204) {
        return null;
    }
    
    // 401 Unauthorized: The token might have expired. Refresh and retry.
    if (response.status === 401) {
        await refreshAccessToken();
        return getCurrentPlaying(); // Retry the request
    }
    
    if (!response.ok) {
        throw new Error(`Spotify API error: ${response.statusText}`);
    }

    return response.json();
}

// This is the main function Netlify will run.
exports.handler = async function(event, context) {
    try {
        const data = await getCurrentPlaying();

        // If nothing is playing or the data is malformed.
        if (!data || !data.item) {
            return {
                statusCode: 200,
                body: JSON.stringify({ playing: false }),
            };
        }

        const item = data.item;
        const result = {
            playing: true,
            track: item.name,
            // .map().join() is the JavaScript equivalent of the Python list comprehension.
            artists: item.artists.map(a => a.name).join(', '),
            albumArt: item.album.images[0].url,
        };

        return {
            statusCode: 200,
            body: JSON.stringify(result),
        };

    } catch (e) {
        // Log the error for debugging in the Netlify function logs.
        console.error(e);
        return {
            statusCode: 500,
            body: JSON.stringify({ playing: false, error: e.toString() }),
        };
    }
};

File: netlify/functions/spotify-current.py
import os
import time
import base64
import requests

access_token = None
token_expires_at = 0

def refresh_access_token():
    global access_token, token_expires_at

    CLIENT_ID = os.environ['SPOTIFY_CLIENT_ID']
    CLIENT_SECRET = os.environ['SPOTIFY_CLIENT_SECRET']
    REFRESH_TOKEN = os.environ['SPOTIFY_REFRESH_TOKEN']

    auth_header = f"{CLIENT_ID}:{CLIENT_SECRET}".encode('utf-8')
    base64_auth = base64.b64encode(auth_header).decode('utf-8')

    response = requests.post(
        'https://accounts.spotify.com/api/token',
        headers={
            'Authorization': f'Basic {base64_auth}',
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        data={
            'grant_type': 'refresh_token',
            'refresh_token': REFRESH_TOKEN
        }
    )

    data = response.json()
    if response.status_code != 200:
        raise Exception(f"Failed to refresh token: {data}")

    access_token = data['access_token']
    token_expires_at = time.time() + data['expires_in']

def get_current_playing():
    global access_token
    if not access_token or time.time() >= token_expires_at:
        refresh_access_token()

    response = requests.get(
        'https://api.spotify.com/v1/me/player/currently-playing',
        headers={'Authorization': f'Bearer {access_token}'}
    )

    if response.status_code == 204:
        return None
    elif response.status_code == 401:
        refresh_access_token()
        return get_current_playing()

    return response.json()

def handler(event, context):
    try:
        data = get_current_playing()
        if not data or 'item' not in data:
            return {
                "statusCode": 200,
                "body": '{"playing": false}'
            }

        item = data['item']
        result = {
            'playing': True,
            'track': item['name'],
            'artists': ', '.join([a['name'] for a in item['artists']]),
            'albumArt': item['album']['images'][0]['url']
        }
        import json
        return {
            "statusCode": 200,
            "body": json.dumps(result)
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "body": '{"playing": false, "error": "%s"}' % str(e)
        }
